### ISSUE TITLE
date parameters should use DateTimeInterface rather than DateTime

### ISSUE THREAD

**Issue by dpash** [jlevers/selling-partner-api#718]
Date parameters currently use `\DateTime` rather than `\DateTimeInterface` meaning that `\DateTimeImmutable` need to be converted to a `\DateTime` first.

(Additionally, the phpdoc is missing the `\` namespace, resulting in Psalm thinking it's a class in the same namespace as the DTO

for example: 

```php
final class CreateReportSpecification extends Dto
{
    /**
     * @param  string  $reportType  The report type. Refer to [Report Type Values](https://developer-docs.amazon.com/sp-api/docs/report-type-values) for more information.
     * @param  string[]  $marketplaceIds  A list of marketplace identifiers for the report.
     * @param  ?string[]  $reportOptions  Additional information passed to reports. This varies by report type.
     * @param  ?DateTime  $dataStartTime  The start of a date and time range, in ISO 8601 date time format, used for selecting the data to report. The default is now. The value must be prior to or equal to the current date and time. Not all report types make use of this.
     * @param  ?DateTime  $dataEndTime  The end of a date and time range, in ISO 8601 date time format, used for selecting the data to report. The default is now. The value must be prior to or equal to the current date and time. Not all report types make use of this.
     */
    public function __construct(
        public readonly string $reportType,
        public readonly array $marketplaceIds,
        public readonly ?array $reportOptions = null,
        public readonly ?\DateTime $dataStartTime = null,
        public readonly ?\DateTime $dataEndTime = null,
    ) {
    }
}
```

should be:

```php
final class CreateReportSpecification extends Dto
{
    /**
     * @param  string  $reportType  The report type. Refer to [Report Type Values](https://developer-docs.amazon.com/sp-api/docs/report-type-values) for more information.
     * @param  string[]  $marketplaceIds  A list of marketplace identifiers for the report.
     * @param  ?string[]  $reportOptions  Additional information passed to reports. This varies by report type.
     * @param  ?\DateTimeInterface  $dataStartTime  The start of a date and time range, in ISO 8601 date time format, used for selecting the data to report. The default is now. The value must be prior to or equal to the current date and time. Not all report types make use of this.
     * @param  ?\DateTimeInterface  $dataEndTime  The end of a date and time range, in ISO 8601 date time format, used for selecting the data to report. The default is now. The value must be prior to or equal to the current date and time. Not all report types make use of this.
     */
    public function __construct(
        public readonly string $reportType,
        public readonly array $marketplaceIds,
        public readonly ?array $reportOptions = null,
        public readonly ?\DateTimeInterface $dataStartTime = null,
        public readonly ?\DateTimeInterface $dataEndTime = null,
    ) {
    }
}
```

I have tried to modify saloon-sdk-generator but I've struggled to build the test schemas:

```diff
diff --git a/src/Generators/DtoGenerator.php b/src/Generators/DtoGenerator.php
index aab7f14..1185988 100644
--- a/src/Generators/DtoGenerator.php
+++ b/src/Generators/DtoGenerator.php
@@ -97,6 +97,11 @@ class DtoGenerator extends BaseDtoGenerator
             $safeType = Utils::isBuiltinType($additionalProperties->type)
                 ? $additionalProperties->type
                 : NameHelper::safeClassName($additionalProperties->type);
+
+            if ($safeType === 'DateTime') {
+                $safeType = '\DateTimeInterface';
+            }
+
             $classType->addProperty('additionalProperties')
                 ->setPublic()
                 ->addComment("@var {$safeType}[]")
diff --git a/src/Helpers/MethodGeneratorHelper.php b/src/Helpers/MethodGeneratorHelper.php
index 650ba8a..843f681 100644
--- a/src/Helpers/MethodGeneratorHelper.php
+++ b/src/Helpers/MethodGeneratorHelper.php
@@ -59,7 +59,7 @@ class MethodGeneratorHelper
 
         $type = $parameter->type;
         if ($type === 'DateTime') {
-            $type = '\DateTime';
+            $type = '\DateTimeInterface';
         } elseif (! Utils::isBuiltInType($type)) {
             if ($namespace === null) {
                 throw new InvalidArgumentException('$namespace must be passed if the type is not a built-in.');
diff --git a/src/Traits/HasArrayableAttributes.php b/src/Traits/HasArrayableAttributes.php
index e7355d3..e82faad 100644
--- a/src/Traits/HasArrayableAttributes.php
+++ b/src/Traits/HasArrayableAttributes.php
@@ -6,6 +6,7 @@ namespace Crescat\SaloonSdkGenerator\Traits;
 
 use Crescat\SaloonSdkGenerator\Exceptions\InvalidAttributeTypeException;
 use DateTime;
+use DateTimeInterface;
 use ReflectionClass;
 
 trait HasArrayableAttributes
@@ -59,7 +60,7 @@ trait HasArrayableAttributes
     {
         if (is_null($value)) {
             return null;
-        } elseif ($value instanceof DateTime) {
+        } elseif ($value instanceof DateTimeInterface) {
             return $value->format(DateTime::RFC3339);
         } elseif (is_string($type)) {
             if (class_exists($type)) {
```

**Comment by jlevers** [comment]
Re: the test schemas – I (unfortunately) haven't updated the tests at all since forking the original `crescatio/saloon-sdk-generator`. Sorry for the confusion there.

**Comment by jlevers** [comment]
Released in `v7.0.0`.
